sw/host/tests/chip/rv_dm/src/jtag_tap_sel.rs

```diff
@@ -42,54 +42,103 @@ fn test_jtag_tap_sel(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
     let lc_state =
         DifLcCtrlState::from_redundant_encoding(jtag.read_lc_ctrl_reg(&LcCtrlReg::LcState)?)?;
 
+    // For test_unlocked* and RMA state, test that we can switch TAP without reset.
+    let (needs_reset, rv_dm_enabled) = match lc_state {
+        DifLcCtrlState::TestUnlocked0
+        | DifLcCtrlState::TestUnlocked1
+        | DifLcCtrlState::TestUnlocked2
+        | DifLcCtrlState::TestUnlocked3
+        | DifLcCtrlState::TestUnlocked4
+        | DifLcCtrlState::TestUnlocked5
+        | DifLcCtrlState::TestUnlocked6
+        | DifLcCtrlState::TestUnlocked7
+        | DifLcCtrlState::Rma => (false, true),
+        DifLcCtrlState::Dev => (true, true),
+        _ => (true, false),
+    };
+
     jtag.disconnect()?;
     transport.pin_strapping("PINMUX_TAP_LC")?.remove()?;
 
     // Test RISC-V TAP
     transport.pin_strapping("PINMUX_TAP_RISCV")?.apply()?;
-    transport.reset_target(opts.init.bootstrap.options.reset_delay, true)?;
+    if needs_reset {
+        transport.reset_target(opts.init.bootstrap.options.reset_delay, true)?;
+    }
 
-    let mut jtag = opts
+    let jtag = opts
         .init
         .jtag_params
         .create(transport)?
-        .connect(JtagTap::RiscvTap)?;
-    let mut lc_state_rv = 0;
-    jtag.read_memory32(
-        top_earlgrey::LC_CTRL_BASE_ADDR as u32 + LcCtrlReg::LcState as u32,
-        core::slice::from_mut(&mut lc_state_rv),
-    )?;
-    let lc_state_rv = DifLcCtrlState::from_redundant_encoding(lc_state_rv)?;
-    assert_eq!(lc_state, lc_state_rv);
+        .connect(JtagTap::RiscvTap);
+
+    assert_eq!(rv_dm_enabled, jtag.is_ok());
+
+    if let Ok(mut jtag) = jtag {
+        let mut lc_state_rv = 0;
+        jtag.read_memory32(
+            top_earlgrey::LC_CTRL_BASE_ADDR as u32 + LcCtrlReg::LcState as u32,
+            core::slice::from_mut(&mut lc_state_rv),
+        )?;
+        let lc_state_rv = DifLcCtrlState::from_redundant_encoding(lc_state_rv)?;
+        assert_eq!(lc_state, lc_state_rv);
+        jtag.disconnect()?;
+    }
 
-    jtag.disconnect()?;
     transport.pin_strapping("PINMUX_TAP_RISCV")?.remove()?;
 
     // Test DFT TAP
     transport.pin_strapping("PINMUX_TAP_DFT")?.apply()?;
-    transport.reset_target(opts.init.bootstrap.options.reset_delay, true)?;
-
-    // For DFT test we need to do raw OpenOCD command.
+    if needs_reset {
+        transport.reset_target(opts.init.bootstrap.options.reset_delay, true)?;
+    }
+
+    // If RV-DM is disabled, strapping DFT tap will result in LC TAP being selected.
```

Comment:
Why is there a link between selecting the DFT tap and rv_dm being enabled? I thought DFT and RV_DM used different strappings and also are not enabled in the same LC states (DFT is not enabled in DV but RV_DM is)

Reply:
DFT is encoded as strapping both LC and RV_DM strapping pins. The pinmux gates rv_dm access by gating on the strapping pin, instead of first use the pin to determine which tap to access and then gate the tap.

Arguably the pinmux logic could be improved but I am not sure if RTL is still changeable?

Reply:
Hum, I thought the strapping pins were just a 2-bit encoded value (so 4 possible values). The [documentation](https://opentitan.org/book/hw/ip/pinmux/doc/theory_of_operation.html?highlight=strap#strap-sampling-and-tap-isolation) does not suggest that each pin is associated with a particular tap.
If the RTL cannot be changed (likely) then the documentation probably should be.

Reply:
Yes, I think this will probably require a documentation update. The behavior of selecting LC TAP by default is probably ok when `TAP_STRAP = 2'b11`. 

@timothytrippel, what is the behavior you have observed wrt the `DFT_STRAP` pins. Do they override the `TAP_STRAP` pins? Or are we just always pulling the `TAP_STRAP` pins by default in the transport? From inspecting the transport configuration it seems that the default `TAP_STRAP = 2'b00`, which makes me think that `DFT_STRAP` override `TAP_STRAP`.

CC @andreaskurth